tests/dns.rs

```diff
@@ -0,0 +1,110 @@
+use std::{
+    io::{BufRead, BufReader, Write},
+    net::TcpStream,
+    process::{Child, Command},
+    sync::{
+        mpsc,
+        mpsc::{Receiver, Sender},
+    },
+    thread,
+    time::Duration,
+};
+
+fn start_server() -> (Child, Receiver<String>) {
+    let (log_sender, log_receiver): (Sender<String>, Receiver<String>) = mpsc::channel();
+    let mut directoryd_process = Command::new("./target/debug/directoryd")
+        .stdout(std::process::Stdio::piped())
+        .spawn()
+        .unwrap();
+
+    let stdout = directoryd_process.stdout.take().unwrap();
+    let sender = log_sender.clone();
+    thread::spawn(move || {
+        let reader = BufReader::new(stdout);
+        reader.lines().map_while(Result::ok).for_each(|line| {
+            sender.send(line).unwrap_or_else(|e| {
+                println!("Failed to send log: {}", e);
+            });
+        });
+    });
+    (directoryd_process, log_receiver)
+}
+
+fn wait_for_server_start(log_receiver: &Receiver<String>) {
+    let mut server_started = false;
+    while let Ok(log_message) = log_receiver.recv_timeout(Duration::from_secs(5)) {
+        if log_message.contains("RPC socket binding successful") {
+            server_started = true;
+            break;
+        }
+    }
+    assert!(
+        server_started,
+        "Server did not start within the expected time"
+    );
+}
+
+fn send_addresses(addresses: &[&str]) {
+    for address in addresses {
+        let mut stream = TcpStream::connect(("127.0.0.1", 8080)).unwrap();
+        let request = format!("POST {}\n", address);
+        stream.write_all(request.as_bytes()).unwrap();
+    }
```

Comment:
This can be done in a followup PR later. This is not super crucial as directory won't be run by regular users.

Reply:
This seems fine to me, as it may show users that multiple clients are connecting to the directory server to send their data. This is also how we handle it in the maker, where it sends its credentials to the directory server and then closes the connection, rather than keeping it open.